### DISCUSSION TITLE
I want to write files in jshook to upload to collections how to do it

### DISCUSSION THREAD

**Discussion by duolabmeng6** [pocketbase/pocketbase#5362] [Q&A]
I would like to implement a custom route `/up` to accomplish the collection of data from uploads in collections.
But there are problems that can't be solved
```
routerAdd("POST", "/up", async (c) => {
    const uploadedFile = c.formFile("files"); 
    console.log("files", JSON.stringify(uploadedFile))
    if (!uploadedFile) {
        throw new BadRequestError("No file uploaded.");
    }
    const formData = new FormData();
    formData.append("files", uploadedFile);
    const uploadResponse = await $http.send({
        url: `http://127.0.0.1:8090/api/collections/uploads/records`,
        method: "POST",
        body: formData,
        headers: {
        },
    });

    return c.json(200, uploadResponse.json);
});
```
```
{
    "collectionId": "xkmbylx1rqoclch",
    "collectionName": "uploads",
    "created": "2024-08-16 04:16:53.065Z",
    "files": [],
    "id": "huvuep9xumalxer",
    "updated": "2024-08-16 04:16:53.065Z"
}
```
The above code has no effect, it won't upload the file.

If it is correct it is the following data
```
{
    "collectionId": "xkmbylx1rqoclch",
    "collectionName": "uploads",
    "created": "2024-08-16 03:49:03.527Z",
    "files": [
        "snipaste_2023_11_03_23_00_55_qz1TScDwV1.png"
    ],
    "id": "hm2swidm285an05",
    "updated": "2024-08-16 03:49:03.527Z"
}
```

Can this be done in a js hook?

**Comment by ganigeorgiev** [comment]
It is not clear what is the purpose of the custom route nor the intended behavior.

Please don't treat the issues tracker and discussions as llm chat and try to elaborate a little more on your use case to understand the bigger picture and what actually you are trying to do (_see [XY problem](https://xyproblem.info/)_).

In general for uploading files programmatically at the moment you have to use the `RecordUpsertForm` as shown in [the docs example](https://pocketbase.io/docs/js-records/#create-new-record-with-data-validations) and in https://github.com/pocketbase/pocketbase/discussions/4956.

_Side-note: Programmatically operating with files is already simplified with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) and will be available with the next v0.23.0 release in the near future (there are no ETAs yet)._

**Reply by duolabmeng6** [reply]
Thank you so much love your app I've solved it!
I'm trying to create a generalized management upload interface and simple file management functionality The application you've developed is perfect for this.

```
/// <reference path="../pb_data/types.d.ts" />
// pb_hooks/main.pb.js

routerAdd("POST", "/up", async (c) => {
    const appUrl = $app.settings().meta.appUrl
    // console.log("files", JSON.stringify(files))
    const formData = new FormData();
    for (let file of c.multipartForm().file.files) {
        const files = $filesystem.fileFromMultipart(file)
        formData.append("files", files);
    }

    const uploadResponse = await $http.send({
        url: `${appUrl}/api/collections/uploads/records`,
        method: "POST",
        body: formData,
        headers: {
        },
    });

    if (uploadResponse.statusCode == 200) {
        const imageUrls = uploadResponse.json.files.map(file => `/api/files/${uploadResponse.json.collectionId}/${uploadResponse.json.id}/${file}`);
        return c.json(200, {
            'uploaded': true,
            'success': 1,           // 0 表示上传失败，1 表示上传成功
            'message': "图片上传成功",
            'url': imageUrls
        });
    }

    return c.json(200, {
        'uploaded': false,
        'success': 0,           // 0 表示上传失败，1 表示上传成功
        'message': "图片上传失败",
        'error': uploadResponse.json.data.files,
        'url': null
    });

});
```

**Reply by ganigeorgiev** [reply]
The JS `pb_hooks` allows direct access to the filesystem and database.
I don't recommend sending internal http requests to the existing web APIs form the hooks, it is inefficient and honestly doesn't make sense to me.

By saying _"I'm trying to create a generalized management upload interface and simple file management functionality"_ doesn't really clarify anything. Why don't you use the existing web APIs? Why the above needs to be executed in a JS hook?

People constantly ask questions and never bother to elaborate what actually they are trying to do. But I'm getting tired and don't want to deal with this type of support anymore.

**Comment by duolabmeng6** [comment]
Thank you very much for your help. I have modified the program based on the documents and API you provided. Thank you very much.
Since I want to be compatible with my old file upload interface developed in php, I need to modify it in this way without using the existing apis.
```
/// <reference path="../pb_data/types.d.ts" />
// pb_hooks/main.pb.js

routerAdd("POST", "/up", async (c) => {
    const collection = $app.dao().findCollectionByNameOrId("uploads")
    const record = new Record(collection)
    const form = new RecordUpsertForm($app, record)
    let formData = []
    for (let file of c.multipartForm().file.files) {
        formData.push($filesystem.fileFromMultipart(file));
    }
    form.addFiles("files", ...formData)
    try {
        form.submit()
    } catch (e) {
        return c.json(200, {
            'uploaded': false,
            'success': 0,
            'message': "图片上传失败",
            'error': e,
            'url': null
        });
    }
    // console.log(JSON.stringify(record))
    const imageUrls = record.get("files").map(file => `/api/files/${record.collectionId}/${record.id}/${file}`);
    return c.json(200, {
        'uploaded': true,
        'success': 1,
        'message': "图片上传成功",
        'url': imageUrls
    });
});
```
